### ISSUE TITLE
CVODE solver: different results on different platforms?!

### ISSUE THREAD

**Issue by agarny** [opencor/opencor#2254]
If you open and run [[OpenCOR]/models/tests/sedml/lorenz.sedml](https://raw.githubusercontent.com/opencor/opencor/master/models/tests/sedml/lorenz.sedml). On Windows, Linux and macOS, you get:

![Windows01](https://user-images.githubusercontent.com/602265/70488109-82958680-1b5c-11ea-9ced-521eaf78add2.png)
![Linux01](https://user-images.githubusercontent.com/602265/70488116-84f7e080-1b5c-11ea-82b6-4e20e62c723c.png)
![macOS01](https://user-images.githubusercontent.com/602265/70488122-87f2d100-1b5c-11ea-8f27-563ae23a5918.png)

At this stage, it all looks visually the same. However, if you run the model one more time, clear the results and rerun it again (i.e. 3rd run), you get:

![Windows03](https://user-images.githubusercontent.com/602265/70488180-bf617d80-1b5c-11ea-8a3d-857831ff05a6.png)
![Linux03](https://user-images.githubusercontent.com/602265/70488182-c1c3d780-1b5c-11ea-86c3-f187df70dcf1.png)
![macOS03](https://user-images.githubusercontent.com/602265/70488183-c4263180-1b5c-11ea-89a5-f28856c4f463.png)

Then, we can see that the end of the top trace is different on Windows, Linux and macOS! Yet, they all use the same model and same solver!

No idea why we are getting different results. Could that be an issue with CVODE itself (everything works as expected using, say, Forward Euler)? I highly doubt it, but then what could be the reason for those discrepancies?!

**Comment by dbrnz** [comment]
The Lorenz equations **are** describing a chaotic system after all...

I'm wondering if there is some subtle rounding going on, possibly in a system supplied intrinsic.

Doesn't seeing different results after the third run not imply that the initial parameter values are different at the start of the third run? And are these initial values not simply the results at the end of the second run? Is it possible to create SEDML that shows the differences after just the one run??

**Comment by agarny** [comment]
> The Lorenz equations **are** describing a chaotic system after all...

True, but... it's not like we are solving it for different parameter values and/or initial conditions. So, I would expect the same results.

> I'm wondering if there is some subtle rounding going on, possibly in a system supplied intrinsic.

Yes, that's what I am thinking.

> Doesn't seeing different results after the third run not imply that the initial parameter values are different at the start of the third run?

If you look at the third run, you can see that everything looks the same until near the end where, all of a sudden, the solution is different on our different platforms.

> And are these initial values not simply the results at the end of the second run?

The initial values of the third run should indeed be the results at the end of the second run.

> Is it possible to create SEDML that shows the differences after just the one run??

I have played the simulation duration quite a bit and have yet to manage to get different results after one run.

Anyway, I clearly need to look into it in more details, which I will do once I am done with issue #2178.

**Comment by agarny** [comment]
FWIW, as part of my work on issue #2178, I test the models we ship with OpenCOR (i.e. the [Hodgkin-Huxley 1952](https://github.com/opencor/opencor/blob/master/models/hodgkin_huxley_squid_axon_model_1952.cellml), [Noble 1962](https://github.com/opencor/opencor/blob/master/models/noble_model_1962.cellml) and [van der Pol 1928](https://github.com/opencor/opencor/blob/master/models/van_der_pol_model_1928.cellml) models) and... even for the first two (especially the N62 model), I had to round the results to two significant digits to ensure the same results on our supported platforms (one run using CVODE).

**Comment by agarny** [comment]
Different tests (on macOS using a release version of OpenCOR):
- I exported the results of [[OpenCOR]/models/tests/sedml/lorenz.sedml](https://github.com/opencor/opencor/blob/master/models/tests/sedml/lorenz.sedml) to a CSV file and compared the results to our [Python-related tests](https://github.com/opencor/opencor/blob/master/src/plugins/support/PythonSupport/tests/data/macos/release/basictests.out) and they basically match. (They are not 100% the same most probably because in one case the printing was done through C++ and in the other through Python.)
- I ran the SED-ML file using a CVODE `Maximum step` of `0`, `0.001` and `0.000001` (from top to bottom):
![CVODE](https://user-images.githubusercontent.com/602265/70881562-d9560100-2031-11ea-8aef-3e1e3ff04a28.png)
- Similarly, I ran the SED-ML file using a Forward Euler `Step` of `1`, `0.001`, `0.000001` and `0.000000001` (from top to bottom):
![FE](https://user-images.githubusercontent.com/602265/70881648-115d4400-2032-11ea-8c7d-ca869abf0527.png)
Bottom line: the Lorenz model is super sensitive. So, I guess I am "happy" with the CVODE solver "giving" different results on different platforms... and therefore closing this issue.
@dbrnz and @nickerso: do you agree with my assessment?
